src/core/lib/security/credentials/tls/grpc_tls_crl_provider.cc

```diff
@@ -53,6 +64,103 @@ std::string IssuerFromCrl(X509_CRL* crl) {
   return ret;
 }
 
+absl::StatusOr<std::shared_ptr<Crl>> ReadCrlFromFile(
+    absl::string_view crl_path) {
+  grpc_slice crl_slice = grpc_empty_slice();
+  grpc_error_handle err = grpc_load_file(crl_path.data(), 1, &crl_slice);
+  if (!err.ok()) {
+    gpr_log(GPR_ERROR, "Error reading file %s", err.message().data());
+    grpc_slice_unref(crl_slice);
+    return absl::InvalidArgumentError("Could not load file");
+  }
+  absl::StatusOr<std::unique_ptr<Crl>> crl =
+      Crl::Parse(std::string(StringViewFromSlice(crl_slice)));
+  if (!crl.ok()) {
+    grpc_slice_unref(crl_slice);
+    return absl::InvalidArgumentError(absl::StrCat(
+        "Parsing crl string failed with result ", crl.status().ToString()));
+  }
+  grpc_slice_unref(crl_slice);
+  return crl;
+}
+
+struct FileData {
+  std::string path;
+  off_t size;
+};
+
+#if defined(GPR_LINUX) || defined(GPR_ANDROID) || defined(GPR_FREEBSD) || \
+    defined(GPR_APPLE)
+std::string GetAbsoluteFilePath(absl::string_view valid_file_dir,
+                                absl::string_view file_entry_name) {
+  return absl::StrFormat("%s/%s", valid_file_dir, file_entry_name);
+}
+
+std::vector<FileData> GetFilesInDirectory(
+    const std::string& crl_directory_path) {
+  DIR* crl_directory;
+  if ((crl_directory = opendir(crl_directory_path.c_str())) == nullptr) {
+    // TODO(gtcooke94) - directory has gone bad, what should we do? Status
+    // return here?
+    return std::vector<FileData>();
+  }
+  std::vector<FileData> crl_files;
+  struct dirent* directory_entry;
+  while ((directory_entry = readdir(crl_directory)) != nullptr) {
+    const char* file_name = directory_entry->d_name;
```

Comment:
Should we add a defensive check to ensure that `file_name` is not nullptr? And if so, we can just continue?